qutebrowser/completion/models/completionmodel.py

```diff
@@ -0,0 +1,234 @@
+            index: The QModelIndex to get item flags for.
+
+        Return: The item data, or None on an invalid index.
+        """
+        if not index.isValid() or role != Qt.DisplayRole:
+            return None
+        if not index.parent().isValid():
+            if index.column() == 0:
+                return self._categories[index.row()].name
+        else:
+            cat = self._categories[index.parent().row()]
+            idx = cat.index(index.row(), index.column())
+            return cat.data(idx)
+
+    def flags(self, index):
+        """Return the item flags for index.
+
+        Override QAbstractItemModel::flags.
+
+        Return: The item flags, or Qt.NoItemFlags on error.
+        """
+        if not index.isValid():
+            return Qt.NoItemFlags
+        if index.parent().isValid():
+            # item
+            return (Qt.ItemIsEnabled | Qt.ItemIsSelectable |
+                    Qt.ItemNeverHasChildren)
+        else:
+            # category
+            return Qt.NoItemFlags
+
+    def index(self, row, col, parent=QModelIndex()):
+        """Get an index into the model.
+
+        Override QAbstractItemModel::index.
+
+        Return: A QModelIndex.
+        """
+        if (row < 0 or row >= self.rowCount(parent) or
+                col < 0 or col >= self.columnCount(parent)):
+            return QModelIndex()
+        if parent.isValid():
+            if parent.column() != 0:
+                return QModelIndex()
+            # store a pointer to the parent category in internalPointer
+            return self.createIndex(row, col, self._categories[parent.row()])
+        return self.createIndex(row, col, None)
+
+    def parent(self, index):
+        """Get an index to the parent of the given index.
+
+        Override QAbstractItemModel::parent.
+
+        Args:
+            index: The QModelIndex to get the parent index for.
+        """
+        parent_cat = index.internalPointer()
+        if not parent_cat:
+            # categories have no parent
+            return QModelIndex()
+        row = self._categories.index(parent_cat)
+        return self.createIndex(row, 0, None)
+
+    def rowCount(self, parent=QModelIndex()):
+        """Override QAbstractItemModel::rowCount."""
+        if not parent.isValid():
+            # top-level
+            return len(self._categories)
+        cat = self._cat_from_idx(parent)
+        if not cat or parent.column() != 0:
+            # item or nonzero category column (only first col has children)
+            return 0
+        else:
+            # category
+            return cat.rowCount()
+
+    def columnCount(self, parent=QModelIndex()):
+        """Override QAbstractItemModel::columnCount."""
+        # pylint: disable=unused-argument
+        return 3
+
+    def canFetchMore(self, parent):
+        """Override to forward the call to the categories."""
+        cat = self._cat_from_idx(parent)
+        if cat:
+            return cat.canFetchMore(parent)
```

Comment:
In 4be3eeb85508d42f77147b3779f7a20e4f1e6515 you added `parent` here - that seems odd. Wouldn't the underlying category model now get a parent which belongs to another model (the `CompletionModel`)? Shouldn't it never get a parent, as from the category model's point of view, all items are toplevel items (i.e. it's a table)?